modules/stochastic_tools/include/surrogates/SurrogateTrainer.h

```diff
@@ -70,11 +85,201 @@ SurrogateTrainer::declareModelData(const std::string & data_name, const T & valu
+  dof_id_type _local_row;
+
+private:
+  /*
+   * Called at the beginning of execute() to make sure values are set properly
+   */
+  void checkIntegrity() const;
+
+  /// Sampler data for the current row
+  std::vector<Real> _row_data;
+
+  /// Vector of reporter names and their corresponding values (to be filled by getTrainingData)
+  std::unordered_map<ReporterName, std::shared_ptr<TrainingDataBase>> _training_data;
+};
+
+template <typename T>
+const T &
+SurrogateTrainer::getTrainingData(const std::string & param)
+{
+  return getTrainingData<T>(getParam<ReporterName>(param));
+}
+
+template <typename T>
+std::vector<const T *>
+SurrogateTrainer::getTrainingDataVector(const std::string & param, bool use_sampler_default)
+{
+  std::vector<const T *> vals;
+  if (isParamValid(param) || !use_sampler_default)
+  {
+    for (const ReporterName & rname : getParam<std::vector<ReporterName>>(param))
+      vals.push_back(&getTrainingData<T>(rname));
+  }
+  else
+  {
+    if (!std::is_same<T, Real>::value)
+      mooseError("Training data values must be type Real if taken from sampler.");
+    for (const Real & col : _row_data)
+      vals.push_back(&col);
+  }
+
+  return vals;
+}
+
+template <typename T>
+const T &
+SurrogateTrainer::getTrainingData(const ReporterName & rname)
+{
+  if (rname.getObjectName() == "sampler")
+  {
+    if (!std::is_same<T, Real>::value)
+      mooseError("Training data values must be type Real if taken from sampler.");
+
+    std::size_t ind = rname.getValueName().find("_");
+    if (ind != std::string::npos)
+    {
+      unsigned int col = MooseUtils::stringToInteger(rname.getValueName().substr(ind + 1), true);
+      if (col >= _sampler.getNumberOfCols())
+        mooseError("Training data from sampler ",
+                   rname,
+                   " is requesting an index greater than the number of sampler columns (",
+                   _sampler.getNumberOfCols(),
+                   ").");
+      return _row_data[col];
+    }
+    else
+      mooseError("Predictor data ", rname, " from sampler is in the wrong format.");
+  }
+  else
+  {
+    auto it = _training_data.find(rname);
+    if (it != _training_data.end())
+    {
+      auto data = std::dynamic_pointer_cast<TrainingData<T>>(it->second);
+      if (!data)
+        mooseError("Reporter value ", rname, " already exists but is of different type.");
+      return data->get();
+    }
+    else
+    {
+      const std::vector<T> & rval = getReporterValueByName<std::vector<T>>(rname);
+      _training_data[rname] = std::make_shared<TrainingData<T>>(rval);
+      return std::dynamic_pointer_cast<TrainingData<T>>(_training_data[rname])->get();
+    }
+  }
+}
+
+class TrainingDataBase
+{
+public:
+  TrainingDataBase() : _is_distributed(false) {}
+
+  virtual ~TrainingDataBase() = default;
+
+  virtual dof_id_type size() const = 0;
+  virtual void setCurrentIndex(dof_id_type index) = 0;
+  bool & isDistributed() { return _is_distributed; }
```

Comment:
```suggestion
  bool isDistributed() const { return _is_distributed; }
```

Reply:
Based on my previous reply, if this was const, then we would need another function like `setDistributed(bool is_distributed)`, but I wanted to keep this class as simple as possible. Let me know if making this const is a better way to go.

Reply:
No, it is fine.